### PR TITLE
feat(datasource/npm): massage non compliant npm repo strings 

### LINKED ISSUE #4610: Massage non-compliant npm repository strings
For cases such as discussed in https://github.com/renovatebot/renovate/issues/4607#issuecomment-539445676, I think we can safely massage this ourselves to the correct format before returning the result from the npm datasource.

### PR DESCRIPTION
**Pull request by olegkrivtsov**
## Changes:

Massage non-complaint npm repository strings #4610 

## Context:

Fixes #4610 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/npm/get.ts [hunk 1]
```diff
@@ -93,6 +94,27 @@ export async function getDependency(
...
       dep.sourceDirectory = res.repository.directory;
     }
+
+    // Massage the repository URL for non-complaint strings (see issue #4610)
+    // Remove the non-complaint segments of path, so the URL looks like "<scheme>://<domain>/<vendor>/<repo>"
+    // and add directory to the repository
+    const sourceUrlCopy = `${sourceUrl}`;
+    const sourceUrlSplit: string[] = sourceUrlCopy.split('/');
+
+    if (sourceUrlSplit.length > 5) {
```
Comment (rarkins) [inline lib/datasource/npm/get.ts:104]: Is this only done for github.com or for all hosts?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/datasource/npm/get.ts (hunk 1 above)
Comment (olegkrivtsov) [reply lib/datasource/npm/get.ts:104]: I think for all hosts

File: lib/datasource/npm/get.ts [hunk 2]
```diff
@@ -93,6 +94,27 @@ export async function getDependency(
...
       dep.sourceDirectory = res.repository.directory;
     }
+
+    // Massage the repository URL for non-compliant strings (see issue #4610)
+    // Remove the non-compliant segments of path, so the URL looks like "<scheme>://<domain>/<vendor>/<repo>"
+    // and add directory to the repository
+    const sourceUrlCopy = `${sourceUrl}`;
+    const sourceUrlSplit: string[] = sourceUrlCopy.split('/');
+
+    if (sourceUrlSplit.length > 7) {
```
Comment (rarkins) [inline lib/datasource/npm/get.ts:104]: I don't think we can be sure that this is a GitHub repo at this point. This massage should only be for github for now so it needs an extra check

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/datasource/npm/get.ts (hunk 2 above)
Comment (olegkrivtsov) [reply lib/datasource/npm/get.ts]: Hi @rarkins that's a good idea. I've added the check:

```
 if (sourceUrlSplit.length > 7 && sourceUrlSplit[2] === 'github.com') {
      /* istanbul ignore next */
      if (dep.sourceDirectory) {
        logger.debug(
          { dependency: packageName },
          `Ambiguity: dependency has the repository URL path and repository/directory set at once; have to override repository/directory`
        );
      }
      dep.sourceUrl = sourceUrlSplit.slice(0, 5).join('/');
      dep.sourceDirectory = sourceUrlSplit
        .slice(7, sourceUrlSplit.length)
        .join('/');
    }
```

Also I added a test for checking that only github repos are massaged.

**Review by rarkins — APPROVED** [review]

File: lib/datasource/npm/get.ts [hunk 3]
```diff
@@ -93,6 +94,27 @@ export async function getDependency(
...
     }
+
+    // Massage the repository URL for non-compliant strings for github (see issue #4610)
+    // Remove the non-compliant segments of path, so the URL looks like "<scheme>://<domain>/<vendor>/<repo>"
+    // and add directory to the repository
+    const sourceUrlCopy = `${sourceUrl}`;
+    const sourceUrlSplit: string[] = sourceUrlCopy.split('/');
+
+    if (sourceUrlSplit.length > 7 && sourceUrlSplit[2] === 'github.com') {
+      /* istanbul ignore next */
```
Comment (rarkins) [inline lib/datasource/npm/get.ts:105]: Please add a test for this instead of ignore

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/datasource/npm/get.ts (hunk 3 above)
Comment (olegkrivtsov) [reply lib/datasource/npm/get.ts:105]: Hi @rarkins I've spent some time figuring out how to get the logger output for using in my test, but no luck for now. Could you please give me some clue/advice how you typically test such code blocks without using `istanbul ignore next`?

Comment (rarkins) [reply lib/datasource/npm/get.ts:105]: It's not necessary that you check the logger output. Just make sure that the code is passed once you remove the istanbul. i.e. the confirmation of the test will be the code coverage, not any direct test.

Comment (olegkrivtsov) [reply lib/datasource/npm/get.ts]: Thank you, done.

**Review by rarkins — APPROVED** [review]

**Pull request merged**